innovation_sweet_spots/pipeline/pilot/investment_predictions/utils.py

```diff
@@ -0,0 +1,323 @@
+    that occur after the start date, if not, the values are set
+    to 0.
+
+    Args:
+        cb_data: Dataframe to have column added to
+        start_date: Date to check if variable occured after
+        variable: Variable to check dates for
+
+    Returns:
+        Dataframe with additional flag indicating if variable
+        happened in the future
+    """
+    cols_to_loop = cb_data.columns[cb_data.columns.str.contains(variable)]
+    cb_data[f"future_{variable}"] = (
+        pd.DataFrame(
+            [
+                (pd.to_datetime(cb_data[col]) > start_date).astype("int")
+                for col in cols_to_loop
+            ]
+        )
+        .transpose()
+        .max(axis=1)
+    )
+    return cb_data
+
+
+def add_n_funding_rounds_in_window(
+    cb_data: pd.DataFrame, start_date: pd.DatetimeIndex, end_date: pd.DatetimeIndex
+) -> pd.DataFrame:
+    """Adds column for how many funding rounds occured in the date window
+
+    Args:
+        cb_data: Dataframe to add number funding rounds data to
+        start_date: Start date of the time window
+        end_date: End date of the time window
+
+    Returns:
+        Dataframe with column added for how many funding rounds occured
+        in the date window
+    """
+    cols_to_loop = cb_data.columns[cb_data.columns.str.contains("funding_round_date")]
+    cb_data["n_funding_rounds_in_window"] = (
+        pd.DataFrame(
+            [
+                pd.to_datetime(cb_data[col])
+                .between(start_date, end_date, inclusive="both")
+                .astype("int")
+                for col in cols_to_loop
+            ]
+        )
+        .transpose()
+        .sum(axis=1)
+    )
+    return cb_data
+
+
+def add_n_month_since_last_investment_in_window(
+    cb_data: pd.DataFrame, start_date: pd.DatetimeIndex, end_date: pd.DatetimeIndex
+) -> pd.DataFrame:
+    """Add column for number of months since last investment in the time window
+
+    Args:
+        cb_data: Dataframe to add number of months since last investment to
+        start_date: Start date of the time window
+        end_date: End date of the time window
+
+    Returns:
+        Dataframe with column added for number of months since last investment
+        in the time window
+    """
+    cols_to_loop = cb_data.columns[cb_data.columns.str.contains("funding_round_date")]
+    latest_investment_window_date = (
+        cb_data[cols_to_loop][
+            pd.DataFrame(
+                [
+                    pd.to_datetime(cb_data[col]).between(
+                        start_date, end_date, inclusive="both"
+                    )
+                    for col in cols_to_loop
+                ]
+            ).transpose()
+        ]
+        .astype("datetime64[ns]")
+        .max(axis=1)
+    )
+    cb_data["n_months_since_last_investment_in_window"] = (
+        ((end_date - latest_investment_window_date) / np.timedelta64(1, "M"))
+        .fillna(-1)
```

Comment:
Why filling in the nulls with `-1` ? Does it mean that there was no investment?

Reply:
Yes, it means no prior investment.